### ISSUE TITLE
Is it possible that (gumshoe--jump) only jumps to buffers in the current window and not across windows?

### ISSUE THREAD

**Issue by Dima-369** [Overdr0ne/gumshoe#9]
I often have 2 windows open side by side and with the backtrack commands, my cursor ends up jumping from left to right so much that it really annoys me. What would work best for me if only the currently open buffer in the window is where the buffer changes take place. Could this maybe be implemented with a `defcustom`?

I tried to modify `gumshoe-display-buffer-action` but with no luck. I also tried to replace `(pop-to-buffer)` with `(switch-to-buffer)` in `(gumshoe--jump)`, but that threw some errors although it was an improvement for me as it works in some cases.

---

I already added this block of code into `gumshoe--jump` which is better in keeping track of the cursor, but I still have to visually track if the point ended up in another window.

```emacs-lisp
(run-at-time nil nil
             (lambda ()
               ;; display longer beacon
               (let ((beacon-size 60)
                     (beacon-blink-delay 0.09)
                     (beacon-color
                      (face-attribute 'font-lock-keyword-face :foreground)))
                 (beacon-blink))))
```

**Comment by Overdr0ne** [comment]
Hey Dima, sorry for the delay, so gumshoe has 3 different scopes. `gumshoe-backtrack-back/forward` operates globally, `gumshoe-buf-backtrack-back/forward` operates only within the buffer, and, if enabled, `gumshoe-persp-backtrack-back/forward` operates only within the current perspective. Similar for the `gumshoe-peruse-in-*` commands. Hope that helps, lmk if you have any more issues.

**Comment by Dima-369** [comment]
Yes, I am indeed aware of the scopes. Let me rephrase:

I often have 2 windows open side by side and with the `gumshoe-backtrack-back/forward`  commands, my cursor ends up jumping from the left to the right window so much that it really annoys me as I always have to keep track to which window it jumps to.

I wondered if there is a way to customize `gumshoe` so it does not switch between windows.

**Comment by Overdr0ne** [comment]
Hmm, I think I see. You want to use global scope, but don't want to jump between windows. Interesting. I think the proper solution would be to implement another scope restricted to windows. It's certainly possible. Other solutions I can think of would be fragile hacks at best.

Also, I do really like your idea of flashing the cursor when it jumps. I think I will steal that one if you don't mind. Though maybe for jumps between windows I could flash the background for the entire window as well. 

I'll play around with things a bit and update when I have something.

Thank you very much for bringing this up.

**Comment by Overdr0ne** [comment]
So if you pull the latest version, you should be able to use the `gumshoe-win-backtrack-*` and `gumshoe-peruse-in-window` commands. And you should see the cursor flash on all jumps. Lmk if that doesn't work for you- otherwise I'll close this issue in the next couple days.

**Comment by Dima-369** [comment]
I tested the `gumshoe-win-backtrack-*` commands, but they never jump for me? But it might be some part of my config. Now, if it works for you great, no need to investigate :)

---

I settled on a different workflow which uses `winner-undo` from the builtin `winner.el` to jump to previous buffers. And then when I found the buffer I wanted, I use something like your `gumshoe-buf-backtrack-*` commands to navigate inside the buffer. I think I had this workflow half-baked in mind when I wrote this, as the requested `gumshoe-win-backtrack-*` would not entirely replicate this behavior. Anyway, thanks for your implementation!

> And you should see the cursor flash on all jumps.

That's nice! I just prefer to use `beacon-blink`, but that's just me.

**Comment by landakram** [comment]
FWIW, I just tried out `gumshoe-win-backtrack-*` functions and they did jump for me. But, they didn't do what I wanted exactly, which I'll try to explain. It might be different than what @Dima-369 was hoping for. 

It seems that the win-backtrack functions jump to any _gumshoe location that was previously open in that window_. What I was hoping for instead is that I could jump to any gumshoe location _globally_, but the jump would occur in the _window where point currently is_, rather than moving point to a different window.

The use-case is that I often go into "explore mode" where I'm opening up different windows side-by-side to explore some call-stack. For example, if the call stack is 4 levels deep, I might even split 4 windows so that I can see the full context of the call-stack. But I also might close some windows so I just see two levels of the call-stack side-by-side. Once I've found what I'm looking for, I generally close _all_ the windows except one and then expect I can backtrack using that one window. Instead, backtracking moves point between windows or even pops windows back up.

I'm pretty sure that the behavior that I want can be achieved by using gumshoe's "global" scope, but with `switch-to-buffer` instead of `pop-to-buffer`, like @Dima-369 mentioned in the original comment

**Comment by landakram** [comment]
Opened a PR that I believe creates the desired behavior and fixes this issue: https://github.com/Overdr0ne/gumshoe/pull/10

The desired behavior can be achieved with `(setq gumshoe-prefer-same-window t)`.

**Comment by smorris-witekio** [comment]
Yeah, I didn't really think deeply about windowing when I was writing all this, because in most cases you can customize windowing behavior with the [zen of buffer display](https://www.gnu.org/software/emacs/manual/html_node/elisp/The-Zen-of-Buffer-Display.html) , but in this case it needs to be customizable by the package directly.

I suspect the reason the window-local gumshoe mode didn't work for you @Dima-369 , is because I had to modify the log structure, which requires the package to be reloaded, because the gumshoe methods will find entries of incompatible types in the log. So restarting could very well fix your issue, but in any case, I hope between that and @landakram 's patch, your problem is resolved. Just lmk if it doesn't and I'll reopen this issue.

Thanks very much again @landakram :pray:

**Comment by Overdr0ne** [comment]
Yeah, I didn't really think deeply about windowing when I was writing all this, because in most cases you can customize windowing behavior with the [zen of buffer display](https://www.gnu.org/software/emacs/manual/html_node/elisp/The-Zen-of-Buffer-Display.html) , but in this case it needs to be customizable by the package directly.

I suspect the reason the window-local gumshoe mode didn't work for you @Dima-369 , is because I had to modify the log structure, which requires the package to be reloaded, because the gumshoe methods will find entries of incompatible types in the log. So restarting could very well fix your issue- I should probably fix that for other people updating. In any case, I hope between that and @landakram 's patch, your problem is resolved. Just lmk if it doesn't and I'll reopen this issue.

Thanks very much again @landakram :pray:
